### PR TITLE
Added support for CloudflareR2 Storage adaptar

### PR DESCRIPTION
**Pull request by Tushar98644**
fixes [appwrite/appwrite#3352](https://github.com/appwrite/appwrite/issues/3352)

Screenshots : 

Screenshot 2023-10-19 at 3 35 33 PM

### CODE REVIEW

**Comment by Tushar98644** [comment]
@stnguyen90 kindly review

**Comment by Tushar98644** [comment]
> @stnguyen90 @gewenyu99
 kindly review

**Comment by bannu0snake** [comment]
hi @Tushar98644 , can you please let me know how did you setup php and testing env in your mac. any reference links ?

**Comment by Tushar98644** [comment]
> hi @Tushar98644 , can you please let me know how did you setup php and testing env in your mac. any reference links ?

I just installed PHP and Composer,after installation you can add the project dependencies using `composer install`. You can have a look at the `composer.json` file , there you can find the scripts which used can be used to run the lining operations . For linting a file simply use `composer format <file-path>` and `composer lint <file-path>` to run the lining test

File: src/Storage/Device/CloudflareR2.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    const ENAM = 'us-east-1';
+
+    const WEUR = 'eu-west-1';
+
+    const EEUR = 'eu-east-1';
+
+    const APAC = 'ap-southeast-1';
+
+    const AUTO = 'auto';
```
Comment (stnguyen90) [inline src/Storage/Device/CloudflareR2.php:26]: Can you link to the reference for this?

File: src/Storage/Device/CloudflareR2.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+     * @param  string  $accessKey
+     * @param  string  $secretKey
+     * @param  string  $bucket
+     * @param  string  $region
+     * @param  string  $acl
+     */
+    public function __construct(string $root, string $accessKey, string $secretKey, string $bucket, string $region = self::APAC, string $acl = self::ACL_PRIVATE)
+    {
+        parent::__construct($root, $accessKey, $secretKey, $bucket, $region, $acl);
+        $this->headers['host'] = $bucket.'.'.'s3'.'.'.$region.'.cloudflarestorage.com';
```
Comment (stnguyen90) [inline src/Storage/Device/CloudflareR2.php:41]: Do you have any documentation on the host?

File: src/Storage/Device/CloudflareR2.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    const WEUR = 'eu-west-1';
+
+    const EEUR = 'eu-east-1';
+
+    const APAC = 'ap-southeast-1';
+
+    const AUTO = 'auto';
+
+    /**
+     * CloudflareR2 Constructor
```
Comment (stnguyen90) [inline src/Storage/Device/CloudflareR2.php:29]:
```suggestion
     * Cloudflare R2 Constructor
```

File: src/Storage/Storage.php [hunk 4]
```diff
@@ -21,6 +21,8 @@ class Storage
 
     const DEVICE_LINODE = 'linode';
 
+    const CLOUDFARE_R2 = 'cloudflare-r2';
```
Comment (stnguyen90) [inline src/Storage/Storage.php:24]: We don't use dashes for any of the others (DO Spaces) so let's not do it here.

```suggestion
    const CLOUDFLARE_R2 = 'cloudflarer2';
```

File: tests/Storage/Device/CloudfareR2Test.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
+<?php
```
Comment (stnguyen90) [inline tests/Storage/Device/CloudfareR2Test.php:1]: file name should be `CloudflareR2Test.php`

File: .github/workflows/tests.yml [hunk 6]
```diff
@@ -75,6 +75,8 @@ jobs:
           WASABI_SECRET: ${{ secrets.WASABI_SECRET }}
           BACKBLAZE_ACCESS_KEY: ${{ secrets.BACKBLAZE_ACCESS_KEY }}
           BACKBLAZE_SECRET: ${{ secrets.BACKBLAZE_SECRET }}
+          CLOUDFLARE_R2_ACCESS_KEY: ${{ secrets.CLOUDFLARE_R2_ACCESS_KEY }}
```
Comment (stnguyen90) [inline .github/workflows/tests.yml:78]: CloudflareR2Test needs to be added above in the matrix too.

File: tests/Storage/Device/CloudfareR2Test.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace Utopia\Tests\Storage\Device;
+
+use Utopia\Storage\Device\CloudflareR2;
+use Utopia\Tests\Storage\S3Base;
+
+class CloudfareR2Test extends S3Base
```
Comment (stnguyen90) [inline tests/Storage/Device/CloudfareR2Test.php:8]:
```suggestion
class CloudflareR2Test extends S3Base
```

**Review by stnguyen90 — CHANGES_REQUESTED** [review]
Great PR! 🤯 We left some comments during the review, please check them out.

Please also make sure to run the test locally and attach a screenshot of the result.

File: src/Storage/Device/CloudflareR2.php (hunk 1 above)
Comment (Tushar98644) [reply src/Storage/Device/CloudflareR2.php:26]: https://developers.cloudflare.com/durable-objects/platform/data-location/       -- I went through this doc

**Comment by Tushar98644** [comment]
> Great PR! 🤯 We left some comments during the review, please check them out.
> 
> Please also make sure to run the test locally and attach a screenshot of the result.

Do you mean the linting tests? I have attached a screenshot for the same in the pr description

**Comment by Tushar98644** [comment]
@stnguyen90 any updates ?

File: src/Storage/Device/CloudflareR2.php (hunk 1 above)
Comment (stnguyen90) [reply src/Storage/Device/CloudflareR2.php:26]: Please add the link as a comment in the code.

File: .github/workflows/tests.yml [hunk 8]
```diff
@@ -58,7 +58,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        devices: [BackblazeTest, DOSpacesTest, LinodeTest, LocalTest, S3Test, WasabiTest]
+        devices: [BackblazeTest, DOSpacesTest, LinodeTest, LocalTest, S3Test, WasabiTest , CloudflareR2Test]
```
Comment (stnguyen90) [inline .github/workflows/tests.yml:61]:
```suggestion
        devices: [BackblazeTest, DOSpacesTest, LinodeTest, LocalTest, S3Test, WasabiTest, CloudflareR2Test]
```

**Review by stnguyen90 — CHANGES_REQUESTED** [review]
Some minor changes.

Please also address https://github.com/utopia-php/storage/pull/97/files#r1370466673 

> Do you mean the linting tests? I have attached a screenshot for the same in the pr description

No, I mean actual tests. Something that confirms this code actually works with Cloudflare R2 storage.

**Comment by Tushar98644** [comment]
> Some minor changes.
> 
> Please also address https://github.com/utopia-php/storage/pull/97/files#r1370466673
> 
> > Do you mean the linting tests? I have attached a screenshot for the same in the pr description
> 
> No, I mean actual tests. Something that confirms this code actually works with Cloudflare R2 storage.

Screenshot 2023-10-31 at 9 44 56 PM

I am having some issues with the card details and adding the r2 storage subscription to my account , would really appreciate it if you could test it from your side

**Comment by gewenyu99** [comment]
Hey,

Due to time constraints, I'm going to mark this PR `hacktoberfest-accepted` for now so you get DO's Hacktoberfest rewards. We'll continue to work with you on this issue for review and merge.

When it is merged, we'll contact you for Appwrite-specific Hacktoberfest swag.

Thanks for helping us improve Appwrite!

**Comment by stnguyen90** [comment]
> I am having some issues with the card details and adding the r2 storage subscription to my account , would really appreciate it if you could test it from your side

How were you validating anything if you didn't have access to Cloudflare R2 storage?

**Review by stnguyen90 — CHANGES_REQUESTED** [review]

**Comment by Tushar98644** [comment]
> > I am having some issues with the card details and adding the r2 storage subscription to my account , would really appreciate it if you could test it from your side
> 
> How were you validating anything if you didn't have access to Cloudflare R2 storage?

I wasn't , i just followed the guide for adding a new adapter

**Comment by docimin** [comment]
Any update?

File: src/Storage/Device/CloudflareR2.php (hunk 2 above)
Comment (docimin) [reply src/Storage/Device/CloudflareR2.php:42]: An S3 API URL looks like this:
https://accountid.r2.cloudflarestorage.com/bucketname

Where account id is your CF Account ID, like this:
https://905fa3fe223dcff4bf6b46911342bcfd.r2.cloudflarestorage.com/test

If you are using the EU Juristiction, you need to add .eu before the .r2, like this:
https://905fa3fe223dcff4bf6b46911342bcfd.eu.r2.cloudflarestorage.com/test

**Comment by docimin** [comment]
If anyone needs R2 to test it out, hit me up if they don't have a license. (It's free however, especially for testing.)

**Comment by stnguyen90** [comment]
Closing as this doesn't actually work.

**Pull request closed without merging**